types/helpers.d.ts

```diff
@@ -1,69 +1,100 @@
+  <M extends Mutations = Mutations, Map extends BaseMethodMap<Commit<M>> = BaseMethodMap<Commit<M>>>(map: Map): { [K in keyof Map]: Method<any> };
 }
 
-interface FunctionMapperWithNamespace<F, R> {
-  (
-    namespace: string,
-    map: Dictionary<(this: CustomVue, fn: F, ...args: any[]) => any>
-  ): Dictionary<R>;
+interface RootMapMutations<Mutations> extends MapMutations<Mutations> {
+  <M extends Mutations = Mutations, Key extends keyof M = keyof M>(namespace: string, map: Key[]): { [K in Key]: MutationMethod<M[K]> };
+  <M extends Mutations = Mutations, Map extends Record<string, keyof M> = Record<string, keyof M>>(namespace: string, map: Map): { [K in keyof Map]: MutationMethod<M[Map[K]]> };
+  <M extends Mutations = Mutations, Map extends BaseMethodMap<Commit<M>> = BaseMethodMap<Commit<M>>>(namespace: string, map: Map): { [K in keyof Map]: Method<any> };
 }
 
-interface MapperForState {
-  <S>(
-    map: Dictionary<(this: CustomVue, state: S, getters: any) => any>
-  ): Dictionary<Computed>;
+/**
+ * mapActions
+ */
+interface MapActions<Actions> {
+  <A extends Actions = Actions, Key extends keyof A = keyof A>(map: Key[]): { [K in Key]: ActionMethod<A[K]> };
+  <A extends Actions = Actions, Map extends Record<string, keyof A> = Record<string, keyof A>>(map: Map): { [K in keyof Map]: ActionMethod<A[Map[K]]> };
+  <A extends Actions = Actions, Map extends BaseMethodMap<Dispatch<A>> = BaseMethodMap<Dispatch<A>>>(map: Map): { [K in keyof Map]: Method<any> };
 }
 
-interface MapperForStateWithNamespace {
-  <S>(
-    namespace: string,
-    map: Dictionary<(this: CustomVue, state: S, getters: any) => any>
-  ): Dictionary<Computed>;
+interface RootMapActions<Actions> extends MapActions<Actions> {
+  <A extends Actions = Actions, Key extends keyof A = keyof A>(namespace: string, map: Key[]): { [K in Key]: ActionMethod<A[K]> };
+  <A extends Actions = Actions, Map extends Record<string, keyof A> = Record<string, keyof A>>(namespace: string, map: Map): { [K in keyof Map]: ActionMethod<A[Map[K]]> };
+  <A extends Actions = Actions, Map extends BaseMethodMap<Dispatch<A>> = BaseMethodMap<Dispatch<A>>>(namespace: string, map: Map): { [K in keyof Map]: Method<any> };
 }
 
-interface NamespacedMappers {
-  mapState: Mapper<Computed> & MapperForState;
-  mapMutations: Mapper<MutationMethod> & FunctionMapper<Commit, MutationMethod>;
-  mapGetters: Mapper<Computed>;
-  mapActions: Mapper<ActionMethod> & FunctionMapper<Dispatch, ActionMethod>;
+/**
+ * namespaced helpers
+ */
+interface NamespacedMappers<State, Getters, Mutations, Actions> {
+  mapState: MapState<State, Getters>;
+  mapGetters: MapGetters<Getters>;
+  mapMutations: MapMutations<Mutations>;
+  mapActions: MapActions<Actions>;
 }
 
-export declare const mapState: Mapper<Computed>
-  & MapperWithNamespace<Computed>
-  & MapperForState
-  & MapperForStateWithNamespace;
+export declare const mapState: RootMapState<BaseType, BaseType>;
```

Comment:
The proposed usage is `mapState<RootState>(['foo', 'bar'])`, however, the return type is `RootState` because `foo/bar` isn't participating inference.

Another usage is that we export  `RootMapXXX` and slightly change their type.

```ts
export interface RootMapState<State> {
  <Keys extends keyof State>(keys: Keys[]): {[K in Keys]: State[K]}
}
```

then users can write something like `const myMapState: RootMapState<RootState> = mapState`.  Of course, it requires us to maintain more public types, and also looks a little bit bizarre.  But it is more versatile in usage -- I think object style can be supported, and more precise in returning type.

Reply:
How about reusing the pattern of `createNameSpacedHelpers`? Adding one more helper function like `createRootHelpers`? Implementation wise, it just return these root helpers, but it gives better developer experience for both end users and lib maintainers.

Reply:
> How about reusing the pattern of createNameSpacedHelpers

That sounds good idea. I think returning root helpers from `createNamespacedHelpers` with no argument would work. `createRootHelpers` would sound a bit weird for pure JS users.

Reply:
Awesome! Great balance between types and runtime behavior.